packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx

```diff
@@ -61,14 +64,14 @@ function UnforwardedToggleGroupControlAsButtonGroup(
 		if ( previousValue !== groupContext.state ) {
 			onChange( groupContext.state );
 		}
-	}, [ groupContext.state ] );
+	}, [ groupContext.state, previousValue, onChange ] );
 
 	// Sync incoming value with groupContext.state.
 	useUpdateEffect( () => {
 		if ( value !== groupContext.state ) {
 			groupContext.setState( value );
 		}
-	}, [ value ] );
+	}, [ value, groupContext ] );
```

Comment:
Since `groupContext.setState` is effectively `setSelectedValue`, I wonder if we could just swap the call inside the hook to use `setSelectedValue` instead of `groupContext.setState`. That way, ESLint should not complain about a missing dependency?

Reply:
I don't see any problem with that approach! One thing to bear in mind though is that to actually eliminate the the `groupContext` dependency, we'd also need to replace `groupContext.state` with `selectedItem`. At that point, `selectedItem` becomes a dependency, so we've kind of come full circle and are still adding a second dependency.

That said, if we feel that `selectedItem` and `setSelectedItem` are more readable than the current `groupContext` values, I'm happy to make that change! 🙂

Reply:
Interestingly, it looks like at least one of our failing tests on this PR is being caused by adding `selectedValue` (or `groupContext` under the previous implementation) to the dependency array of the second `useUpdateEffect`. 

It doesn't look like the deselection is actually working. The test clicks the button once, then clicks it again, but the `pressed` stated doesn't look like it's updating. I'll keep investigating.

Reply:
ah. yes that makes sense. Thank you!

Changes applied, waiting on CI. 🤞

Reply:
Tested these changes on Storybook — it seems like the `onChange` callback gets called twice every time the value changes 

<img width="964" alt="Screenshot 2022-11-26 at 18 51 11" src="https://user-images.githubusercontent.com/1083581/204102289-c73aa693-e71f-4a70-82f2-74edd583d1f2.png">

I wonder if using [`useControlledValue`](https://github.com/WordPress/gutenberg/blob/5cca2e1a8021679aaf2fc713c1a2c1cce168c374/packages/components/src/utils/hooks/use-controlled-value.ts) instead of the custom logic would help here with syncing controlled and uncontrolled updates?

Reply:
Oh, promising ideal. I've pushed an attempt at this change in [64ef11f](https://github.com/WordPress/gutenberg/pull/45771/commits/64ef11f6bb488e6f4c0a83f09de6fe774fcade60). I had to address a type discrepancy on the setter function, let me know how it looks!

Reply:
I prefer not to have to pass an empty string instead of `undefined`, since that is rather arbitrary and not a solution that can be adopted in the general case (e.g., what if our state was a `number` type?). In fact, it's probably worth fixing this limitation in `useControlledValue` itself (not in this PR).

In our specific case, I feel like the most straightforward fix for the double onChange problem is something like this:

```jsx
	// Propagate selectedValue change.
	useUpdateEffect( () => {
		// Only call onChange when uncontrolled
		if ( previousValue === value ) {
			onChange( selectedValue );
		}
	}, [ selectedValue, previousValue, value, onChange ] );
```

This seems to be working in my brief testing for both controlled/uncontrolled cases, though I think it's worth adding tests. Marco and I were talking about how we might want to start adding a [`.each()`](https://jestjs.io/docs/api#testeachtablename-fn-timeout) to our component tests so everything is tested in both controlled/uncontrolled cases. @chad1008 Do you want to try that here? It might not fit the scope of this PR, and this PR has been running long, so feel free to defer if it's too high-effort to include here. As long as we are reasonably confident that we haven't regressed, this PR can land.